### PR TITLE
Optional KFrag validation in `pre.reencrypt()` (and other stuff)

### LINKED ISSUE #227: pytest-benchmark fails with pytest 4
CircleCI failure: https://circleci.com/gh/nucypher/pyUmbral/2277

See https://github.com/ionelmc/pytest-benchmark/issues/122

### PR DESCRIPTION
**Pull request by cygnusv**
Assorted improvements:
- Validation of KFrags is now optional when re-encrypting (although the default remains to be to validate always). This can be useful in scenarios where the re-encrypting node ("Ursula") has already verified the KFrag (which is a reasonable assumption) and she can be sure that this KFrag hasn't changed.
- Module `umbral.signing` allows to sign and verify prehashed messages.
- Introduces badges in the README.
- Improvement of the CI build time.
- Closes #227.

### CODE REVIEW

File: umbral/dem.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 DEM_NONCE_SIZE = 12
 
 
-class UmbralDEM(object):
+class UmbralDEM():
```
Comment (KPrasch) [inline umbral/dem.py]:
``` suggestion
class UmbralDEM:
```

File: umbral/kfrags.py [hunk 2]
```diff
@@ -127,17 +127,27 @@ def to_bytes(self) -> bytes:
 
     def verify(self,
                signing_pubkey: UmbralPublicKey,
-               delegating_pubkey: UmbralPublicKey = None,
-               receiving_pubkey: UmbralPublicKey = None,
+               delegating_pubkey: Optional[UmbralPublicKey] = None,
```
Comment (KPrasch) [inline umbral/kfrags.py]: :+1:

File: umbral/kfrags.py [hunk 3]
```diff
@@ -127,17 +127,27 @@ def to_bytes(self) -> bytes:
...
+            elif delegating_pubkey.params != params:
+                raise ValueError("The delegating pubkey has different UmbralParameters.")
+
+        if self.receiving_key_in_signature():
+            if receiving_pubkey is None:
+                raise ValueError("The receiving pubkey is required to verify this KFrag.")
+            elif receiving_pubkey.params != params:
+                raise ValueError("The receiving pubkey has different UmbralParameters.")
 
         u = params.u
```
Comment (KPrasch) [inline umbral/kfrags.py]: `u`. so lonely over here....

File: umbral/cfrags.py [hunk 4]
```diff
@@ -29,7 +29,7 @@
 from umbral.random_oracles import hash_to_curvebn, ExtendedKeccak
 
 
-class CorrectnessProof(object):
+class CorrectnessProof():
```
Comment (michwill) [inline umbral/cfrags.py]: I think, you don't need `()`

**Pull request merged**